http-clients/src/main/java/feign/OptionalAwareDecoder.java

```diff
@@ -35,11 +36,19 @@ public OptionalAwareDecoder(Decoder delegate) {
 
     @Override
     public Object decode(Response response, Type type) throws IOException, FeignException {
-        if (response.status() == 204 && Types.getRawType(type).equals(Optional.class)) {
-            return Optional.absent();
+        if (Types.getRawType(type).equals(Optional.class)) {
+            if (response.status() == 204) {
+                return Optional.absent();
+            } else {
+                return Optional.of(delegate.decode(response, getInnerType(type)));
```

Comment:
Throwing probably makes sense, it's basically a programming error. Problem is the NPE is going to be a pain to debug. How about this?

    Optional.of(checkNotNull(delegate.decode(response, getInnerType(type)), "Unexpected null response content for status %d", response.status());